docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,119 @@
+
+- Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4743
+
+## Context and Problem Statement
+
+Now that we have the [new policy matching resources](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/005-policy-matching.md), it's time to create one for traffic routung.
+We currently have both `TrafficRoute` and `MeshGatewayRoute`.
+
+This MADR does not address Gateway API integration.
+
+## Decision Drivers
+
+- `MeshGatewayRoute` and `TrafficRoute` have a large overlap in functionality
+  and the use cases for routing from a gateway and between services are similar
+- Gateway API has different resources per protocol
+
+## Considered Options
+
+- Two resources: `MeshGatewayRoute` and `MeshTrafficRoute`
+- Unified resource for services and gateway: `MeshTrafficRoute`
+- Unified resources but one per protocol, e.g. `MeshHTTPRoute`
+
+## Decision Outcome
+
+The new resource routes and alters requests depending on where it's coming from and where it's going to. Routes end up in the client-side Envoy config so we choose a `spec.to` resource.
+
+```yaml
+spec:
+ targetRef: ... # from where requests are coming
+ to:
+  - targetRef: ... # to where requests are going
+ default: ...
+```
+
+Each protocol gets its own resource:
+
+- `MeshHTTPRoute`
+- `MeshTCPRoute`
+
+which can be extended with others like GRPC.
+
+### Positive Consequences
+
+- Route configuration can be very precisely targeted
+- Users need know only one set of resources for routing
+- A final spec that has the best of `MeshGatewayRoute` and `TrafficRoute`
+
+### New `MeshHTTPRoute` spec
+
+Routing configuration consists of a list of rules where each rule _matches_ some
+requests or traffic, _modifies_ it in some way and then sends it to a _destination_.
+
+#### Matching
+
+The new route matches requests based on:
+
+- headers
+- path
+- method
+- query parameters
+
+A rule contains a list of match objects, each of which specifies a match for one
+or more of the above types:
+
+```yaml
+  matches:
+  - method:
+    path:
+    header:
+    queryParams:
+```
+
+A given entry in `match` succeeds if _all_ of the conditions are satisfied and the rule matches when _any one_ of the entries under `match` suceeds. In other words, `match` is an OR'd list of AND'd conditions.
+
+##### Headers
+
+Headers can be tested for:
+
+- presence
+- absence
+- exact value
+- prefix value
+- regex value
+
+#### Request modification
+
+Aspects that can be modified:
+
+- request headers
+  - host
+- response headers
+- request path
+- HTTP redirect
+- request mirroring (response is ignored)
+
+```yaml
+  modifies:
+  ...
+```
+
+#### Routing
+
+If `match` succeeds, the request is routed to the specified destinations.
+In the case that a request should be routed to the service it would otherwise
+have been routed to, destinations can be omitted.
+
+Otherwise, `destinations` holds a list of weighted destinations for the traffic:
+
+```yaml
+- destinations:
```

Comment:
For these entry is there much reason to diverge from the GatewayAPI spec?

Reply:
No, we should use `targetRef`

Reply:
What I meant in my first comment is should we have the similar object as the GatewayAPI [HTTPRoute](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRouteRule) but with a list of rules that has matches, filters and backendRefs though in our case the `backendRef` would be our TargetRef?

In a way that's very similar to the current MeshGatewayRoute

Reply:
Yeah it's already the same idea of course. Or do you mean we should make it as similar as possible including field names.

Reply:
I think we decided we can't use the:

```
type: prefix|regex
value: ...
```

pattern and instead need

```
prefix: ...
regex: ...
```

right? @lobkovilya

Reply:
Yes I meant including field name. Feels like the gatewayApi folks spent time picking these field name and arch and might not be useful to reinvent the wheel?

I think type is not a problem anymore since we use straight go structs now but I could be wrong.

Reply:
I've made it more like HTTPRoute. One thing I think I want to change from HTTP route is being able to match one of a list of HTTP methods. It seems more concise than only one at a time.

I think they're equally powerful but:

```
rules:
- matches:
  - methods: 
    - GET
    - POST
    path:
      type: Exact
      value: /api
```

seems cleaner than

```
rules:
- matches:
  - method: GET
    path:
      type: Exact
      value: /api
  - method: POST
    path:
      type: Exact
      value: /api
```

Reply:
I'm changing my mind on this, we can always add `methods` later.